Prevent hidden files from being parsed

When using Mac OS X to edit Openhab files through an SMB share, files with a leading ._ are created by Mac OS X to store file information about the actual file in.

**Example:**
Create/edit a file called "test.item" using OS X finder application  and Mac OS X also creates a file called "._test.item".

Although these ._ files do not contain any valid information for Openhab, they are being parsed causing problems in mainly the Openhab rule engine.
This leads to errors in the Openhab log like:

2016-09-02 09:54:45.764 [ERROR] [.script.engine.ScriptExecutionThread] - Rule 'Pantry light motion': Script interpreter couldn't be obtain

2016-09-01 15:47:55.026 [WARN ] [pse.smarthome.core.items.GenericItem] - failed notifying listener '[org.eclipse.smarthome.model.rule.runtime.internal.engine.RuleEngineImpl@1ec05f5, Home_State, null]' about state update of item java.lang.NullPointerException: {}

Perhaps preventing the ._ files to be parsed would help in preventing issues caused by these files?

--- a/bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java
+++ b/bundles/config/org.eclipse.smarthome.config.dispatch/src/main/java/org/eclipse/smarthome/config/dispatch/internal/ConfigDispatcher.java
@@ -159,7 +159,10 @@ public class ConfigDispatcher extends AbstractWatchService {
     protected void processWatchEvent(WatchEvent<?> event, Kind<?> kind, Path path) {
         if (kind == ENTRY_CREATE || kind == ENTRY_MODIFY) {
             try {
-                processConfigFile(path.toFile());
+                File f = path.toFile();
+                if (!f.isHidden()) {
+                    processConfigFile(f);
+                }
             } catch (IOException e) {
                 logger.warn("Could not process config file '{}': {}", path, e);
             }

--- a/bundles/model/org.eclipse.smarthome.model.core.test/src/test/groovy/org/eclipse/smarthome/model/core/internal/folder/test/FolderObserverTest.groovy
+++ b/bundles/model/org.eclipse.smarthome.model.core.test/src/test/groovy/org/eclipse/smarthome/model/core/internal/folder/test/FolderObserverTest.groovy
@@ -10,6 +10,8 @@ package org.eclipse.smarthome.model.core.internal.folder.test
 import static org.hamcrest.CoreMatchers.*
 import static org.junit.Assert.*
 
+import java.nio.file.Files
+
 import org.apache.commons.io.FileUtils
 import org.apache.commons.lang.SystemUtils
 import org.eclipse.emf.ecore.EObject
@@ -42,7 +44,8 @@ import org.osgi.service.cm.ManagedService
  *  Since the assertions in the tests depend on handling the events,
  *  putting the current Thread to sleep after the file operations is also necessary.
  *
- *  @author Mihaela Memova
+ *  @author Mihaela Memova - Initial Contribution
+ *  @author Stefan Triller - added hidden file test
  *
  */
 
@@ -484,6 +487,43 @@ class FolderObserverTest extends OSGiTest {
 
         waitForAssert {
             assertThat modelRepo.isAddOrRefreshModelMethodCalled, is(true)
+        }
+    }
+
+    @Test
+    void 'test ignore hidden files with extension that is registered'() {
+        /*
+         * The following method creates a hidden file in an existing directory. The file's extension is
+         * in the configuration properties and there is a registered ModelParser for it.
+         * addOrRefreshModel() method invocation is NOT expected, the model should be ignored since the file is hidden
+         */
+        String validExtension = "java"
+
+        Dictionary<String, String> configProps = new Hashtable<String, String>()
+        configProps.put(EXISTING_SUBDIR_NAME, "txt,jpg," + validExtension)
+        config.update(configProps)
+        sleep(WAIT_ABSTRACTWATCHQUEUEREADER_TO_START)
+
+        String mockFileWithValidExtName = ".HiddenNewlyCreatedMockFile" + "." + validExtension
+        String mockFileWithValidExtPath = EXISTING_SUBDIR_PATH + File.separatorChar + mockFileWithValidExtName
+        File mockFileWithValidExt = new File(mockFileWithValidExtPath)
+        mockFileWithValidExt.createNewFile()
+        sleep(WAIT_EVENT_TO_BE_HANDLED)
+        /*
+         * In some OS, like MacOS, creating an empty file is not related to sending an ENTRY_CREATE event.
+         * So, it's necessary to put some initial content in that file.
+         */
+        if(!SystemUtils.IS_OS_WINDOWS) {
+            mockFileWithValidExt << INITIAL_FILE_CONTENT
+        }
+        else {
+            //windows needs a filesystem property set for a file to be hidden (untested)
+            Files.set.setAttribute (mockFileWithValidExt.toPath(), "dos:hidden", true);
+        }
+
+        waitForAssert{
+            assertThat("A model should NOT be added/refreshed on new hidden file creation in the watched directory",
+                    modelRepo.isAddOrRefreshModelMethodCalled,is(false))
         }
     }
 

--- a/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
+++ b/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
@@ -311,7 +311,7 @@ public class FolderObserver extends AbstractWatchService implements ManagedServi
     @Override
     protected void processWatchEvent(WatchEvent<?> event, Kind<?> kind, Path path) {
         File toCheck = getFileByFileExtMap(folderFileExtMap, path.getFileName().toString());
-        if (toCheck != null) {
+        if (toCheck != null && !toCheck.isHidden()) {
             checkFile(modelRepo, toCheck, kind);
         }
     }

--- a/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
+++ b/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
@@ -236,7 +236,8 @@ public class FolderObserver extends AbstractWatchService implements ManagedServi
             try {
                 synchronized (FolderObserver.class) {
                     if ((kind == ENTRY_CREATE || kind == ENTRY_MODIFY)) {
-                        if (parsers.contains(getExtension(file.getName()))) {
+                        // we omit parsing of hidden files possibly created by editors or operating systems
+                        if (parsers.contains(getExtension(file.getName())) && !file.getName().startsWith(".")) {
                             try (FileInputStream inputStream = FileUtils.openInputStream(file)) {
                                 modelRepo.addOrRefreshModel(file.getName(), inputStream);
                             } catch (IOException e) {

--- a/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
+++ b/bundles/model/org.eclipse.smarthome.model.core/src/main/java/org/eclipse/smarthome/model/core/internal/folder/FolderObserver.java
@@ -165,7 +165,10 @@ public class FolderObserver extends AbstractWatchService implements ManagedServi
                     File[] files = folder.listFiles(new FileExtensionsFilter(validExtension));
                     if (files != null && files.length > 0) {
                         for (File file : files) {
-                            checkFile(modelRepo, file, ENTRY_CREATE);
+                            // we omit parsing of hidden files possibly created by editors or operating systems
+                            if (!file.isHidden()) {
+                                checkFile(modelRepo, file, ENTRY_CREATE);
+                            }
                         }
                     }
                 }
@@ -236,8 +239,7 @@ public class FolderObserver extends AbstractWatchService implements ManagedServi
             try {
                 synchronized (FolderObserver.class) {
                     if ((kind == ENTRY_CREATE || kind == ENTRY_MODIFY)) {
-                        // we omit parsing of hidden files possibly created by editors or operating systems
-                        if (parsers.contains(getExtension(file.getName())) && !file.getName().startsWith(".")) {
+                        if (parsers.contains(getExtension(file.getName()))) {
                             try (FileInputStream inputStream = FileUtils.openInputStream(file)) {
                                 modelRepo.addOrRefreshModel(file.getName(), inputStream);
                             } catch (IOException e) {

